### ISSUE TITLE
Allow connecting to an in-memory dummy database

### ISSUE THREAD

**Issue by tobiasdiez** [prisma/prisma#11129]
## Problem

Currently the documentation for testing either recommends mocking PrismaClient or a full integration testing.
However, in many scenarios, having a database around just for integration testing is overkill and adds additional overhead (e.g. one has to handle database migration etc). 

## Suggested solution

Create a "mock" of PrismaClient that operates completely on an in-memory cache, without writing anything to a real database. In this way one could easily create a test environment for light-weight integration tests e.g. testing the front-end with some predefined dummy data (without mocking the complete backend).

## Alternatives

Use a database that has an in-memory mode, such as SQLLite. However, one still needs to handle migration of the schema etc. Related issue for this https://github.com/prisma/prisma/issues/732.

**Comment by capaj** [comment]
@tobiasdiez thanks for opening this, but I believe this is a duplicate.
If you think about it, having in memory connector without prisma automatically managing the migration for the developer would be tedious, because you'd have to manually trigger the migration every time you spin up your nodejs process.
So if we run in memory, prisma should just always run the migration on startup to get the DB into desired state.
So if I was a maintainer, I would probably close this one as dupe of #732.
Unless you suggest they reimplement the whole sqlite as part of prisma, which sounds out of scope anyway.

**Comment by tobiasdiez** [comment]
My idea was to implement a backend for prisma that stores everything as plain javascript objects, no serialization and no execution of sql. Hence there is also no need to migrate anything since you simply start in an empty state. This would also make development so much easier since one doesn't need to set-up a database.

You can compare this to an in-memory store for sessions, which you would replace with a redis store in production.

**Comment by phorcys420** [comment]
Would especially be useful for unit testing !

**Comment by Toinouze** [comment]
This is crucial for testing. Any updates ?

**Comment by dgosza** [comment]
This would be so helpful to unit tests. Still waiting

**Comment by joaoh9** [comment]
+1

**Comment by empz** [comment]
It's a shame that this is not built-in into Prisma. It seems it would be quite simple to implement an in-memory database. At least, much simpler than implement a database-specific.

For all of you coming here, I'm having good results with the following library: https://github.com/morintd/prismock - I think it's better than prisma-mock.

**Comment by pitazzo** [comment]
this feature would be a killer

**Comment by janpio** [comment]
This could probably be implemented with either Client Extensions, then the code would need to use the Prisma Client queries and turn that into the operation on the local memory cache of what the database should contain, or with driver adapters, where one could use the SQL queries to achieve something similar - possibly with an existing in memory implementation of a SQL database.

**Comment by denis-sokolov** [comment]
Note that adapting Prisma to [PGlite](https://github.com/electric-sql/pglite) might be a path forward.

**Comment by lekhasy** [comment]
This is not just a feature, it's faith.
It's hard to trust someone who consider unit testing as an afterthought.
This is the only thing make me not suggest prisma to new projects.

**Comment by nahtnam** [comment]
https://github.com/stackframe-projects/pgmock Also looks really cool (similar to pg-lite)

**Comment by KaKi87** [comment]
Turns out there is a [Prisma adapter for PGLite](https://github.com/lucasthevenet/pglite-utils/tree/main/packages/prisma-adapter), but there's no way to only enable it during testing. 😭

**Comment by IArnaut2** [comment]
It's stupid that Prisma does not have in-memory database support. In-memory databases are very useful in testing.

**Comment by thw0rted** [comment]
In case it helps anybody, our team was successful at setting up unit testing via pglite on a NestJS application.  We use Nest DI to inject a pglite-based PrismaClient connected to a pglite instance.  At its core the setup function just runs

```
npx prisma migrate diff --from-empty --to-schema ./prisma/schema.prisma --script
```

then passes the resulting SQL to pglite.  Each test file (outer `describe` block) gets its own instance of the client so the specs can be run in parallel, and after each spec we run a cleanup function that's basically

```
prisma.$queryRawUnsafe(`TRUNCATE TABLE "${t}" RESTART IDENTITY CASCADE;`)
```

looping over all tables inside a transaction.

Overall this approach is a bit slower than running a real containerized PG but much much cleaner.  I'd still love to see official support but this worked well for us.

**Comment by electrovir** [comment]
I have a package that almost seamlessly uses PGlite with Prisma. I use this in dev and tests (with a fresh DB per test). It's _much_ faster than PG containers.

https://www.npmjs.com/package/prisma-pglite
